multi/stm32l4-multi/tty.c

```diff
@@ -99,60 +140,179 @@ static int tty_irqHandler(unsigned int n, void *arg)
 {
 	tty_ctx_t *ctx = (tty_ctx_t *)arg;
 
-	if (*(ctx->base + isr) & ((1 << 5) | (1 << 3))) {
+	if ((*(ctx->base + isr) & ((1 << 5) | (1 << 3))) != 0) {
 		/* Clear overrun error bit */
 		*(ctx->base + icr) |= (1 << 3);
 
-		lf_fifo_push(&ctx->rx_fifo, *(ctx->base + rdr));
-		ctx->rxready = 1;
+		lf_fifo_push(&ctx->irq.rx_fifo, *(ctx->base + rdr));
+		ctx->irq.rxready = 1;
 	}
 
-	if (tty_txready(ctx))
+	if (tty_txready(ctx) != 0) {
 		*(ctx->base + cr1) &= ~(1 << 7);
+	}
 
 	return 1;
 }
 
 
+static void tty_dmaCallback(void *arg, int type)
+{
+	tty_ctx_t *ctx = (tty_ctx_t *)arg;
+	unsigned int write_pos = __atomic_load_n(&ctx->dma.write_pos, __ATOMIC_SEQ_CST);
+	unsigned int read_pos = __atomic_load_n(&ctx->dma.read_pos, __ATOMIC_SEQ_CST);
+	unsigned int new_write_pos;
+
+	switch (type) {
+		case dma_ht:
+			new_write_pos = (ctx->dma.rxbufsz / 2);
+			ctx->dma.prev_left_to_rx = (ctx->dma.rxbufsz / 2);
+			break;
+		case dma_tc:
+			new_write_pos = 0;
+			ctx->dma.prev_left_to_rx = ctx->dma.rxbufsz;
+			break;
+		/* Uart idle */
+		default:
+			new_write_pos = (write_pos + (ctx->dma.prev_left_to_rx - (-type))) & ctx->dma.rxbufsz_mask;
+			ctx->dma.prev_left_to_rx = -type;
+			break;
+	}
+
+	__atomic_store_n(&ctx->dma.write_pos, new_write_pos, __ATOMIC_SEQ_CST);
```

Comment:
It would be nice to implement lock-less fifo as a separate module. That would enhance readability of this code and allow perhaps to use this module in different places too